### PR TITLE
Refactor / simplify FCS support.

### PR DESCRIPTION
**Pull request by AlexVlx**
Function call enablement is strictly of interest for the ME and BE, the FE doesn't have much use for this information. Thus, this changes things over by adding a -hc-function-calls driver flag, which the linker knows what to do with (essentially instructs SelectAcceleratorCode whether or not to make all reachable functions `always_inline`).

### CODE REVIEW

File: lib/clamp-device.in [hunk 1]
```diff
@@ -210,18 +211,6 @@ if [ $RETVAL != 0 ]; then
   exit $RETVAL
 fi
 
-# Optimization notes:
-#  -disable-simplify-libcalls:  prevents transforming loops into library calls such as memset, memcopy on GPU
-$OPT -mtriple amdgcn-amd-amdhsa -mcpu=$AMDGPU_TARGET \
-  -amdgpu-internalize-symbols -disable-simplify-libcalls $KMOPTOPT -verify \
```
Comment (scchan) [inline lib/clamp-device.in:216]: The combined `opt` invocation doesn't explicitly specify `amdgpu-internalize-symbols`, does it still run it?

File: tests/Unit/FunctionCall/fcs_saxpy.cpp [hunk 2]
```diff
@@ -1,4 +1,4 @@
-// RUN: %hc -hc -amdgpu-function-calls %s -o %t.out && %t.out
+// RUN: %hc -hc -Wl,-hc-function-calls %s -o %t.out && %t.out
```
Comment (scchan) [inline tests/Unit/FunctionCall/fcs_saxpy.cpp]: `-hc-function-calls` should still work as a compiler switch?  In the use case with early finalize, the linker doesn't get invoked when we generate the isa.

File: .gitmodules [hunk 3]
```diff
@@ -9,7 +9,7 @@
 [submodule "clang"]
 	path = clang
 	url = https://github.com/RadeonOpenCompute/hcc-clang-upgrade.git
-	branch = clang_tot_upgrade
+	branch = feature_refactor_dfcs
```
Comment (scchan) [inline .gitmodules]: This should be changed back to clang_tot_upgrade

**Review by scchan — CHANGES_REQUESTED** [review]

File: tests/Unit/FunctionCall/fcs_saxpy.cpp (hunk 2 above)
Comment (AlexVlx) [reply tests/Unit/FunctionCall/fcs_saxpy.cpp]: Yeah, that's where it'll end up, there's just a piece missing.

File: .gitmodules (hunk 3 above)
Comment (AlexVlx) [reply .gitmodules]: Eventually yes, but for now it allows people to just clone and use it directly, without having to check something out in the submodule.

File: lib/clamp-device.in (hunk 1 above)
Comment (AlexVlx) [reply lib/clamp-device.in:216]: LC already runs that anyway. I think we were just doing work that LC does for no clear reason / for historical reasons lost to time. We should really only run the passes that are specific to us, IMHO, possibly without any optimisation, and then let LC's pipeline take care of everything.

File: .gitmodules (hunk 3 above)
Comment (AlexVlx) [reply .gitmodules]: Ah, I didn't notice you merged the Clang PR, thanks. However, there'll be a follow-up to handle the hc-function-calls not-as-linker-flag scenario that you outlined above.

Comment (david-salinas) [reply .gitmodules]: I don't believe the branch field will make a difference.  The gitref contains the commit id, and the path and url fields give git all it needs to get the submodule.

**Comment by AlexVlx** [comment]
I've injected a bit of a kludge in that we now run `opt` with whatever optimisation level is encoded in KMOPTOPT for the custom HCC passes. Whilst this addresses a pernicious issue, it's somewhat unsatisfactory since we only want to run our specific passes, and defer to LC on all other matters; this'll remain on the TODO list for some follow-up work.

**Review by scchan — APPROVED** [review]
LGTM, thanks!

**Pull request merged**